halo1/model.py

```diff
@@ -115,3 +116,92 @@ def import_halo1_markers(jms, *, scale=1.0, node_size=0.01,
 		set_translation_from_jms(scene_marker, marker, scale)
 
 	#TODO: Should this return something?
+
+def import_halo1_region_from_jms(jms, *, name="", scale=1.0, region_filter=()):
+	'''
+	Imports all the geometry into a Halo 1 JMS into the scene.
+	'''
+
+	if not len(region_filter):
+		region_filter = range(len(jms.regions))
+
+	### Geometry preprocessing.
+
+	# Ready the vertices.
+	vertices = tuple(map(
+			lambda v : (v.pos_x * scale, v.pos_y * scale, v.pos_z * scale),
+			jms.verts
+		)
+	)
+
+	# Ready the triangles.
+
+	# Filter the triangles so only the wished regions are retrieved.
+	triangles = filter(lambda t : t.region in region_filter, jms.tris)
+
+	# Reduce the triangles to just their key components.
+	triangles = tuple(map(lambda t : (t.v0, t.v1, t.v2), triangles))
+
+	# Unpack the vertex normals.
+	vertex_normals = tuple(
+		map(lambda v : (v.norm_i, v.norm_j, v.norm_k), jms.verts)
+	)
+
+	# Convert the vertex normals to triangle normals.
+	tri_normals = map(
+		lambda t : (
+			vertex_normals[t[0]],
+			vertex_normals[t[1]],
+			vertex_normals[t[2]]),
+		triangles
+	)
+
+	# Remove unused vertices
+	vertices, triangles = reduce_vertices(vertices, triangles)
+
+	# Chain all of the triangle normals together into loop normals.
+	# ((x, y, z), (x, y, z), (x, y, z)), ((x, y, z), (x, y, z), (x, y, z)),
+	#    |    |    |
+	#    V    V    V
+	# ((x, y, z), (x, y, z), (x, y, z), (x, y, z), (x, y, z), (x, y, z)),
+	# Loops are the points of triangles so to speak.
+
+	loop_normals = tuple(itertools.chain(*tri_normals))
+
+	### Importing the data into a mesh
+
+	# Make a mesh to hold all relevant data.
+	mesh = bpy.data.meshes.new(name)
+
+	# Import the verts and tris into the mesh.
+	# verts, edges, tris. If () is given for edges Blender will infer them.
+	mesh.from_pydata(vertices, (), triangles)
+
+	# Import loop normals into the mesh.
+	mesh.normals_split_custom_set(loop_normals)
+
+	# Blender will only display custom vertex normals with this enabled.
+	# 0.0 vectors are passed through to use actual auto smooth.
+	mesh.use_auto_smooth = True
```

Comment:
We're not going to autosmooth forever, right? Will we eventually import the smoothing from the model itself?

Reply:
This _is_ importing the smoothing from the model itself. I tried to explain that in the comment.

Reply:
Oh, that is not clear from the comment at all. Is that what "custom vertex normals" means?

Reply:
Yeah! thanks!